Index: .cvsignore
===================================================================
RCS file: /sources/qemu/qemu/.cvsignore,v
retrieving revision 1.23
diff -u -r1.23 .cvsignore
--- .cvsignore	8 Nov 2007 16:43:05 -0000	1.23
+++ .cvsignore	7 Jan 2008 23:41:09 -0000
@@ -1,5 +1,6 @@
 config-host.*
 dyngen
+dyngen.dSYM
 i386
 *-softmmu
 *-darwin-user
Index: Makefile
===================================================================
RCS file: /sources/qemu/qemu/Makefile,v
retrieving revision 1.140
diff -u -r1.140 Makefile
--- Makefile	6 Jan 2008 18:27:12 -0000	1.140
+++ Makefile	7 Jan 2008 23:41:10 -0000
@@ -146,6 +146,7 @@
 # avoid old build problems by removing potentially incorrect old files
 	rm -f config.mak config.h op-i386.h opc-i386.h gen-op-i386.h op-arm.h opc-arm.h gen-op-arm.h
 	rm -f *.o *.d *.a $(TOOLS) dyngen$(EXESUF) TAGS cscope.* *.pod *~ */*~
+	rm -rf dyngen.dSYM
 	rm -f slirp/*.o slirp/*.d audio/*.o audio/*.d
 	$(MAKE) -C tests clean
 	for d in $(TARGET_DIRS); do \
