--- block.c	2007-02-06 02:15:39.000000000 +0100
+++ block2.c	2007-02-06 02:24:10.000000000 +0100
@@ -497,6 +497,7 @@
 int bdrv_read(BlockDriverState *bs, int64_t sector_num, 
               uint8_t *buf, int nb_sectors)
 {
+    bs->activityLED = 1;
     BlockDriver *drv = bs->drv;
 
     if (!drv)
@@ -534,6 +535,7 @@
 int bdrv_write(BlockDriverState *bs, int64_t sector_num, 
                const uint8_t *buf, int nb_sectors)
 {
+    bs->activityLED = 1;
     BlockDriver *drv = bs->drv;
     if (!bs->drv)
         return -ENOMEDIUM;
@@ -1058,6 +1060,7 @@
                                 uint8_t *buf, int nb_sectors,
                                 BlockDriverCompletionFunc *cb, void *opaque)
 {
+    bs->activityLED = 1;
     BlockDriver *drv = bs->drv;
 
     if (!drv)
@@ -1078,6 +1081,7 @@
                                  const uint8_t *buf, int nb_sectors,
                                  BlockDriverCompletionFunc *cb, void *opaque)
 {
+    bs->activityLED = 1;
     BlockDriver *drv = bs->drv;
 
     if (!drv)
