--- block_int.h	2006-06-21 18:46:53.000000000 +0200
+++ block_int2.h	2006-07-03 17:20:04.000000000 +0200
@@ -51,6 +51,7 @@
     int removable; /* if true, the media can be removed */
     int locked;    /* if true, the media cannot temporarily be ejected */
     int encrypted; /* if true, the media is encrypted */
+    int activityLED; /* if true, the media is accessed atm */
     /* event callback when inserting/removing */
     void (*change_cb)(void *opaque);
     void *change_opaque;
