--- pc.c	2006-06-14 18:03:05.000000000 +0200
+++ pc.c_2	2006-07-13 09:07:00.000000000 +0200
@@ -192,16 +192,16 @@
     switch(boot_device) {
     case 'a':
     case 'b':
-        rtc_set_memory(s, 0x3d, 0x01); /* floppy boot */
+        rtc_set_memory(s, 0x3d, 0x21); /* floppy boot */
         if (!fd_bootchk)
             rtc_set_memory(s, 0x38, 0x01); /* disable signature check */
         break;
     default:
     case 'c':
-        rtc_set_memory(s, 0x3d, 0x02); /* hard drive boot */
+        rtc_set_memory(s, 0x3d, 0x32); /* hard drive boot */
         break;
     case 'd':
-        rtc_set_memory(s, 0x3d, 0x03); /* CD-ROM boot */
+        rtc_set_memory(s, 0x3d, 0x23); /* CD-ROM boot */
         break;
     }
 
